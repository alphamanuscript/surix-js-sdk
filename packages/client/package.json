{
  "name": "@surix/client",
  "version": "0.7.2",
  "description": "Surix API Client",
  "keywords": [
    "surix",
    "client",
    "wrapper",
    "api",
    "rest"
  ],
  "author": "Alpha Manuscript <surix@manuscript.live>",
  "homepage": "https://github.com/alphamanuscript/surix-js-sdk#readme",
  "license": "ISC",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "directories": {
    "lib": "src"
  },
  "files": [
    "lib",
    "dist"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/alphamanuscript/surix-js-sdk.git"
  },
  "scripts": {
    "test": "jest src",
    "test:watch": "jest --watch src",
    "test:e2e": "jest test/e2e",
    "test:all": "yarn test",
    "clean": "rm -rf lib dist",
    "build": "yarn clean && yarn build:node && yarn build:umd",
    "build:node": "tsc",
    "build:umd": "webpack --mode=production",
    "prepublish": "yarn test:all && yarn build"
  },
  "bugs": {
    "url": "https://github.com/alphamanuscript/surix-js-sdk/issues"
  },
  "jest": {
    "rootDir": ".",
    "transform": {
      "^.+\\.tsx?$": "ts-jest"
    },
    "testRegex": "(/tests/.*|(\\.|/)(test|spec))\\.(js?|ts?)$",
    "moduleFileExtensions": [
      "ts",
      "js"
    ]
  },
  "devDependencies": {
    "@types/axios": "^0.14.0",
    "@types/jest": "^23.3.9",
    "jest": "^23.6.0",
    "ts-jest": "^23.10.5",
    "tslint": "^5.11.0",
    "typescript": "^3.1.6",
    "uglifyjs-webpack-plugin": "^2.0.1",
    "webpack": "^4.26.0",
    "webpack-cli": "^3.1.2"
  },
  "dependencies": {
    "@surix/data-helpers": "^0.7.1",
    "axios": "^0.18.0"
  }
}
