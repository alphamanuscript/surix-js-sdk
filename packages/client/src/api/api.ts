import axios, { AxiosInstance } from 'axios';

export function getApiClient (baseUrl: string, apiKey: string): AxiosInstance {
  const axiosInstance = axios.create({
    baseURL: baseUrl
  })
  axiosInstance.defaults.headers['Authorization'] = `apikey ${apiKey}`
  return axiosInstance
}