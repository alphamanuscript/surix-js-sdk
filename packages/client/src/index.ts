export { Client } from './client';
export { Project } from './types';
export * from '@surix/data-helpers';
