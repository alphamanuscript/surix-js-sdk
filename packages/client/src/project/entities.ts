import { 
  ApiEntity,  
  expandEntity, 
  expandQuery, 
  wrapEntity, 
  wrapEntityArray, 
  WrappedEntity
} from '@surix/data-helpers';
import { AxiosInstance } from 'axios';
import {  DeletedEntities, Entity, ProjectEntities, Query } from '../types';

export function getProjectEntities (projectId: string, apiClient: AxiosInstance): ProjectEntities {
  return {
    async get(entityId: string): Promise<WrappedEntity> {
      const res = await apiClient.get<ApiEntity>(`/projects/${projectId}/entities/${entityId}`);
      return wrapEntity(res.data);
    },
    async query(query: Query = {}): Promise<WrappedEntity[]> {
      const expandedQuery = expandQuery(query);
      if (Object.keys(expandedQuery.query).length === 0) {
        delete expandedQuery.query;
      }
      const res = await apiClient.post<ApiEntity[]>(
        `/projects/${projectId}/entities/query`, expandedQuery);
      const entities = res.data;
      return wrapEntityArray(entities);
    },
    async update(entity: Entity): Promise<WrappedEntity> {
      const _entity = { ...entity };
      const entityId = _entity._id;
      delete _entity._id;
      const expandedEntity = expandEntity(_entity);
      const res = await apiClient.patch<ApiEntity>(
        `/projects/${projectId}/entities/${entityId}`, expandedEntity);
      const returnedEntity = res.data;
      return wrapEntity(returnedEntity);
    },
    async put(entity: Entity): Promise<WrappedEntity> {
      const _entity = { ...entity }; // Make a shallow copy
      const entityId = _entity._id;
      delete _entity._id;
      const expandedEntity = expandEntity(_entity);
      const res = await apiClient.put<ApiEntity>(
        `/projects/${projectId}/entities/${entityId}`, expandedEntity);
      const returnedEntity = res.data;
      return wrapEntity(returnedEntity);
    },
    async delete(entityId: string): Promise<WrappedEntity> {
      const res = await apiClient.delete(`/projects/${projectId}/entities/${entityId}`);
      const returnedEntity = res.data;
      return wrapEntity(returnedEntity);
    },
    async deleteMany(entityIds: string[]): Promise<DeletedEntities> {
      const res = await apiClient.delete(`/projects/${projectId}/entities`, 
      {
        data: {
          entities: entityIds
        }
      });
      return res.data;
    },
    async create(entity: Entity): Promise<WrappedEntity> {
      const expandedEntity = expandEntity(entity);
      const res = await apiClient.post<ApiEntity>(
        `/projects/${projectId}/entities`, expandedEntity);
      const returnedEntity = res.data;
      return wrapEntity(returnedEntity);
    }
  };
}