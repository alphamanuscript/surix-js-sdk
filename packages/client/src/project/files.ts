import {
  ApiFile,
  wrapFile,
  wrapFileArray,
  WrappedFile } from '@surix/data-helpers';
import { AxiosInstance } from 'axios';
import { ProjectFiles} from '../types';

export function getProjectFiles (projectId: string, apiClient: AxiosInstance): ProjectFiles {
  return {
    async get (fileId: string): Promise<WrappedFile> {
      const res = await apiClient.get<ApiFile>(`/projects/${projectId}/files/${fileId}`);
      return wrapFile(res.data);
    },
    async list (): Promise<WrappedFile[]> {
      const res = await apiClient.get<ApiFile[]>(`/projects/${projectId}/files`);
      return wrapFileArray(res.data);
    }
  };
}