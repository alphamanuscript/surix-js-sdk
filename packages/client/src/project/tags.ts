import { AxiosInstance } from 'axios';
import { ProjectTags, Tag, TagList  } from '../types';

export function getProjectTags (projectId: string, apiClient: AxiosInstance): ProjectTags {
  return {
    async list (): Promise<TagList> {
      const res = await apiClient.get<TagList>(`/projects/${projectId}/tags`);
      return res.data;
    }, 
    async rename (oldName: string, newName: string): Promise<Tag> {
      const res = await apiClient.put<Tag>(`/projects/${projectId}/tags/${oldName}`, {
        name: newName
      });
      return res.data;
    }
  };
}