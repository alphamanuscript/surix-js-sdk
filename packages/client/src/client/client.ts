import { getApiClient } from '../api';
import { getProjectApi } from '../project';
import { Project } from '../types';

export const PRODUCTION_URL = 'https://api.surix.co';
export const STAGING_URL = 'https://surix-staging.herokuapp.com';

export interface ClientOptions {
  environment?: 'production' | 'staging' ;
}

export class Client {
  private _apiClient;

  constructor (opts?: ClientOptions) {
    const _opts = opts || {
      environment: 'production'
    };
    const baseUrl = _opts.environment === 'production' ? PRODUCTION_URL : STAGING_URL;
    this._apiClient = getApiClient(baseUrl);
  }

  project (id: string): Project {
    return getProjectApi(id, this._apiClient);
  }
}



