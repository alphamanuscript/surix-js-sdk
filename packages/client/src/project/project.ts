import { AxiosInstance } from 'axios';
import { getProjectDb } from './db';
import { getProjectFiles } from './files';

import { Project } from '../types';

export function getProjectApi (projectId: string, apiClient: AxiosInstance): Project {
  return {
    db: getProjectDb(projectId, apiClient),
    files: getProjectFiles(projectId, apiClient)
  };
}