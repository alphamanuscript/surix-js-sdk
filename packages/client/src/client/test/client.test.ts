import * as api from '../../api';
import * as projectApi from '../../project';
import { Client } from '../client';

describe('Client', () => {
  describe('new instance', () => {
    it('should create http client based on production Surix API url', () => {
      const spy = jest.spyOn(api, 'getApiClient');
      new Client();
      expect(spy.mock.calls).toMatchSnapshot();
      spy.mockRestore();
    });
    describe('when environment is not set to production', () => {
      it('should create http client based on staging url', () => {
        const spy = jest.spyOn(api, 'getApiClient');
        new Client({ environment: 'staging' });
        expect(spy.mock.calls).toMatchSnapshot();
        spy.mockRestore();
      });
    });
    it('should override baseUrl if baseUrl option is provided', () => {
      const spy = jest.spyOn(api, 'getApiClient');
      const baseUrl = 'http://mydomain.com/api';
      new Client({ baseUrl });
      expect(spy.mock.calls).toMatchSnapshot();
      spy.mockRestore();
    });
  });

  describe('project', () => {
    it('should call getProjectApi with project id and apiClient', () => {
      const apiClient = {};
      jest.spyOn(api, 'getApiClient').mockReturnValue(apiClient);
      jest.spyOn(projectApi, 'getProjectApi');
      const client = new Client();
      client.project('id');
      expect(projectApi.getProjectApi).toHaveBeenCalledWith('id', apiClient);
    });
  });
});