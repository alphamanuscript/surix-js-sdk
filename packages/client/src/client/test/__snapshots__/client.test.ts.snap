// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client new instance should create http client based on production Surix API url 1`] = `
Array [
  Array [
    "https://api.surix.co",
  ],
]
`;

exports[`Client new instance when environment is not set to production should create http client based on staging url 1`] = `
Array [
  Array [
    "https://api.surix.co",
  ],
  Array [
    "https://surix-staging.herokuapp.com",
  ],
]
`;
